climpred/predictability_horizon.py

```diff
@@ -0,0 +1,60 @@
+import numpy as np
+import xarray as xr
+
+
+def _last_item_cond_true(cond, dim, shift=-1):
+    """Get the last item of condition ``cond`` which is True.
+
+    Args:
+        cond (xr.DataArray, xr.Dataset): Condition True means predictable. ``cond``
+            contains at least the dimension lead and checks for how many leads the
+            ``condition`` is true and hence skill is predictable.
+        dim (str): Dimension to check for condition == True over.
+        shift (int): Description of parameter `shift`. Defaults to 1.
+
+    Returns:
+        xr.DataArray, xr.Dataset: ``dim`` value until condition is True.
+
+    """
+    reached = cond.idxmin(dim) + shift  # to get the last True lead
+    # fix below one
+    reached = reached.where(reached >= 1, np.nan)
```

Comment:
I think we actually want `argmin` here instead of `idxmin`.

With `idxmin` it grabs the coordinate label for the lead. So we subtract one to get the last time it was true. But right now you're subtracting one from the coordinate label. What if they have irregular spacing on leads? What if for some reason it's `[0.5, 0.75, 1, ...]` or `[1, 2, 4, 5, 6]`. We oculd either grab the wrong coordinate or error using coordinate labeling.

If we use `argmin` we can just use `isel()` instead of `sel()`.

Reply:
For example, it breaks in this case. I know it's weird, but people do weird things.

```python
import climpred
import esmtools

hind = climpred.tutorial.load_dataset('CESM-DP-SST')
hind = esmtools.stats.rm_trend(hind, dim='init')
verif = climpred.tutorial.load_dataset('ERSST')
verif = esmtools.stats.rm_trend(verif, dim='time')

hindcast = climpred.HindcastEnsemble(hind).add_observations(verif)

result = hindcast.verify(metric='acc', dim='init', comparison='e2o', alignment='maximize', reference='persistence')

cond = (result.sel(skill='initialized') > result.sel(skill='persistence'))
cond = cond.SST
# Change leads to something weird
cond['lead'] = [0.25, 0.75, 1, 3, 4, 5, 6, 7, 8, 9]

# Works (all True, so should be 9)
_last_item_cond_true(cond, 'lead')
>>> 9

# Works (first false, so should be nan)
cond[0] = False
_last_item_cond_true(cond, 'lead')
>>> nan

# Fails (second False, so should be 0.25). 
cond[0] = True
cond[1] = False
_last_item_cond_true(cond, 'lead')
>>> nan

# Fails (starting lead with 0 as in some S2S cases)
cond['lead'] = np.arange(cond.lead.size)
cond[0] = True
cond[1] = False
_last_item_cond_true(cond, 'lead')
>>> nan # (should be 0)

```

Keep in mind we do have people who start with lead 0 for instance as the coordinate label so `reached = reached.where(reached >= 1, np.nan)` would break that.

Reply:
So we just need to find the solution to deal with the fact that

1. `argmin` returns -1 for the case that all are True and should instead return `cond.lead.size - 1`.
2. `argmin` returns -1 for the case that the first is False and should instead return `nan`.

Then we could do some sort of transform to go from index space to coordinate space.

Reply:
> some sort of transform to go from index space to coordinate space.

is easy for a single data point, but I'm struggling to get this one a grid

Reply:
`argmin` works better. integrated this testing framework into the tests. btw: condition from verify was false for last lead when I reproduced.